src/tool/confiscan.sh

```diff
@@ -116,6 +116,68 @@ else
     mkdir -p "${output_dir}"
 fi
 
+####################
+# Users and groups #
+####################
+
+passwd_file="/etc/passwd"
+[[ -f "${passwd_file}" ]] || error "passwd file not found." 1
+[[ -r "${passwd_file}" ]] || error "passwd file not readable." 1
+
+groups_file="/etc/group"
+[[ -f "${groups_file}" ]] || error "group file not found." 1
+[[ -r "${groups_file}" ]] || error "group file not readable." 1
+
+[[ -d "${output_dir}/users" ]] || mkdir -p "${output_dir}/users"
+[[ -d "${output_dir}/groups" ]] || mkdir -p "${output_dir}/groups"
+
+printf 'Username,UID,GID,Shell,Home\n' > "${output_dir}/users/users.csv"
+cut -d: -f1 ${passwd_file} | paste -d, - \
+    <(cut -d: -f3 ${passwd_file}) \
+    <(cut -d: -f4 ${passwd_file}) \
+    <(cut -d: -f7 ${passwd_file}) \
+    <(cut -d: -f6 ${passwd_file}) >> \
+    "${output_dir}/users/users.csv"
+
+info "Users found on the system:"
+column -t -s, "${output_dir}/users/users.csv"
+
+info "Users with login shells:"
+while read -r shell; do
+    [[ "${shell}" =~ ^#|^$ ]] && continue
+
+    info "Users with '${shell}':"
+    # ||: is required since the script will exit with 1
+    # if no users are found with the shell.
+    grep -E ":${shell}$" ${passwd_file} | cut -d: -f1 || :
+done < "/etc/shells"
```

Comment:
Maybe something like:

```suggestion
while read -r shell; do
    info "Users with '${shell}':"
    # ||: is required since the script will exit with 1
    # if no users are found with the shell.
    grep -E ":${shell}$" ${passwd_file} | cut -d: -f1 || :
done < <(grep -E ^/ "/etc/shells")
```

would be better? Then the check for comments and empty lines isn't needed anymore.

Reply:
Great idea to list the users with an actual login shell, however I don't think it's really necessary to list their actual shell?

Maybe something like:

```
[6.638s][~]$ grep -f /etc/shells /etc/passwd                                    <<<
root:x:0:0:root:/root:/bin/bash
bdeferme:x:1000:1000:Bert:/home/bdeferme:/usr/bin/zsh
```

Would suffice?

Reply:
If you want to keep the output by shell,yeah sure filter out comments and empty lines before the loop.

Reply:
Tbh users can already see this in the parsed `passwd` output. So I will probably remove it.

https://github.com/AntonVanAssche/hogent-bachelorproef/blob/08985efe6e0ff9c1013cd1d8b58c9e338722331d/src/tool/confiscan.sh#L148-L155